src/validation.cpp

```diff
@@ -4671,32 +4671,44 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
                     continue;
             } catch (const std::exception&) {
                 // no valid block header found; don't complain
+                // (this happens at the end of every blk.dat file)
                 break;
             }
             try {
-                // read block
+                // read block header
                 uint64_t nBlockPos = blkdat.GetPos();
                 if (dbp)
                     dbp->nPos = nBlockPos;
-                blkdat.SetLimit(nBlockPos + nSize);
-                blkdat.SetPos(nBlockPos);
-                std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
-                CBlock& block = *pblock;
-                blkdat >> block;
-                nRewind = blkdat.GetPos();
-
-                uint256 hash = block.GetHash();
+                CBlockHeader header;
+                blkdat >> header;
+
+                // Read the rest of the block (from the end of its header to the end
+                // of the block) to ensure it's in memory while holding cs_main (below).
+                blkdat.Skip(nBlockPos + nSize - blkdat.GetPos());
```

Comment:
This `Skip()` can be skipped. What the PR does now is:

1. Read from disk and deserialize the block header.
2. Read from disk the remaining of the block with this `Skip()` call (just buffer it into `blkdat`).
3. If it is out-of-order block, continue with the next block. Notice that even if the data we read from disk in 2. remains buffered in `blkdat`, we will never make use of it because we read out-of-order blocks using `ReadBlockFromDisk()` which will not use our `blkdat` cache and will read the block from disk _again_ (2x read).
4. If it is in-order block, deserialize the cached data from 2.

We can do better (find a patch at the bottom of this comment):

1. Read from disk and deserialize the block header.
2. If it is out-of-order block, continue with the next block.
3. If it is in-order block, read from disk and deserialize it.

It follows that we don't need the newly introduced method `Skip()` since we only call it here.

<details>
<summary>proposed change that skips the `Skip()` call (without IO while holding `cs_main`)</summary>

```diff
diff --git i/src/validation.cpp w/src/validation.cpp
index 34a6e028b..705595625 100644
--- i/src/validation.cpp
+++ w/src/validation.cpp
@@ -4657,39 +4657,42 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
                 uint64_t nBlockPos = blkdat.GetPos();
                 if (dbp)
                     dbp->nPos = nBlockPos;
                 CBlockHeader header;
                 blkdat >> header;
 
-                // Read the rest of the block (from the end of its header to the end
-                // of the block) to ensure it's in memory while holding cs_main (below).
-                blkdat.Skip(nBlockPos + nSize - blkdat.GetPos());
-
                 const uint256 hash = header.GetHash();
+                bool outOfOrder;
+
                 {
                     LOCK(cs_main);
-                    // detect out of order blocks, and store them for later
-                    if (hash != chainparams.GetConsensus().hashGenesisBlock && !LookupBlockIndex(header.hashPrevBlock)) {
-                        LogPrint(BCLog::REINDEX, "%s: Out of order block %s, parent %s not known\n", __func__, hash.ToString(),
-                                header.hashPrevBlock.ToString());
-                        if (dbp)
-                            mapBlocksUnknownParent.insert(std::make_pair(header.hashPrevBlock, *dbp));
-
-                        // Skip the rest of this block; position to the marker before the next block.
-                        nRewind = nBlockPos + nSize;
-                        continue;
-                    }
+                    outOfOrder =
+                        hash != chainparams.GetConsensus().hashGenesisBlock
+                        && !LookupBlockIndex(header.hashPrevBlock);
+                }
 
-                    // This block can be processed immediately; rewind to its start then read it.
-                    blkdat.SetPos(nBlockPos);
-                    blkdat.SetLimit(nBlockPos + nSize);
-                    std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
-                    CBlock& block = *pblock;
-                    blkdat >> block;
-                    nRewind = blkdat.GetPos();
+                // Store positions of out of order blocks for later.
+                if (outOfOrder) {
+                    LogPrint(BCLog::REINDEX, "%s: Out of order block %s, parent %s not known\n", __func__, hash.ToString(),
+                            header.hashPrevBlock.ToString());
+                    if (dbp)
+                        mapBlocksUnknownParent.insert(std::make_pair(header.hashPrevBlock, *dbp));
+                    // Skip the rest of this block; position to the marker before the next block.
+                    nRewind = nBlockPos + nSize;
+                    continue;
+                }
+
+                // This block can be processed immediately; rewind to its start then read it.
+                blkdat.SetPos(nBlockPos);
+                std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
+                CBlock& block = *pblock;
+                blkdat >> block;
+                nRewind = blkdat.GetPos();
 
+                {
+                    LOCK(cs_main);
                     // process in case the block isn't known yet
                     CBlockIndex* pindex = LookupBlockIndex(hash);
                     if (!pindex || (pindex->nStatus & BLOCK_HAVE_DATA) == 0) {
                       BlockValidationState state;
                       if (::ChainstateActive().AcceptBlock(pblock, state, chainparams, nullptr, true, dbp, nullptr)) {
                           nLoaded++;
```
</details>

Reply:
Thanks, Vasil. Steps 1 and 2 (in your first list) are not separate disk reads; each disk read generally includes multiple blocks. The `CBufferedFile` object doesn't know (and has no way to know) where the block boundaries are, so it does as large a read as the buffer allows, which is about 4 MB (`MAX_BLOCK_SERIALIZED_SIZE = 4000000`). The buffer boundaries and block boundaries aren't synchronized in any way, so each read usually includes a partial block at the end. (Blocks never span `blk00nnn.dat` files.)

I did consider changing this function to do exactly what you suggest -- read only the block header of each block, and read the rest of the block only if we can accept it immediately. We'd read fewer bytes from disk -- but the tradeoff is that we'd do many more read operations. It's unclear which is better. I suspect it would depend on the underlying kind of disk. I have an SSD, so small reads at different file offsets would probably be efficient, but if it's a spinning disk, doing fewer large reads is probably more efficient since there's less rotational delay.

Also, making that change would be more complex.

Did you time your patch? When your patch sets `nRewind` to the beginning of the next block and takes the `continue` path, that rewind value is used as the argument to `blkdat.SetPos()` at [line 4656](https://github.com/LarryRuane/bitcoin/blob/0e2746148225cbef8954b0c1081e135f6f23a661/src/validation.cpp#L4656) in this PR. I added some debug prints, and as I suspected, that `SetPos()` call isn't always doing what we expect. The caller doesn't check its return value (there's much to be desired about this code); if it can't move ahead to the requested position, it [moves ahead as far as it can](https://github.com/LarryRuane/bitcoin/blob/0e2746148225cbef8954b0c1081e135f6f23a661/src/streams.h#L816) and returns `false`. 

When running your patch, `SetPos()` usually can move to the requested position, but about 30 times per file, it can't because it reaches the end of the buffer (`SetPos` isn't designed to read from disk). We end up usually positioned to the middle of a block, and then [`FindByte()`](https://github.com/LarryRuane/bitcoin/blob/0e2746148225cbef8954b0c1081e135f6f23a661/src/validation.cpp#L4663) searches for (and finds) the start of the next block. It works, but obviously it's not very efficient since each byte is inspected individually.

I think the best way to avoid reading most blocks from disk twice is an idea I described in the last two paragraphs of [this comment](https://github.com/bitcoin/bitcoin/pull/16981#issuecomment-594125403). But that's a much larger change and I don't know if it's worth it. But even if we implement that idea later, I still think this PR, as it is, has a high enough performance improvement to complexity ratio to be merged.

Reply:
Uh, oh. You are right that `SetPos()` must be within what has already been read into the buffer (`CBufferedFile::vchBuf`). I tried timing the suggested patch and it was either the same or ~1 min slower than the PR. This is within noise, but maybe is because of what you describe above (`FindByte()` checking every byte).

`CBufferedFile` is indeed not designed to skip through data (without reading it from disk) and changing it to be able to do so would require some substantial changes and increase in its complexity.

Scrap this suggestion.